### ISSUE TITLE
opencv windows release x86 dll built locally or prebuilt have bugs

### ISSUE THREAD

**Issue by siddhupiddu** [opencv/opencv#7622]
##### System information (version)

- OpenCV => 3.1
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2013

##### Detailed description

OpenCV windows x86 release dlls  built locally using visual studio 2013 exhibit problems that are not in x86 debug dlls. This behaviour is also seen in 2.4.13 both prebuilt dlls and those built locally using visual studio 2013. Debug x86 works fine. Debug and release x64 works fine as well.

##### Steps to reproduce
Download the following base64:
http://pastebin.com/kNM5cDDZ

Load this base64 into a string array called base64image. Decode the encoded string using the method base64_decode using code from here 
http://www.adp-gmbh.ch/cpp/common/base64.html

Then run the following c++ code

<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
		string decoded = base64_decode(base64image);
		vector<uchar> vectordata(decoded.begin(), decoded.start() + decoded.size());
		auto mat = imdecode(vectordata, IMREAD_UNCHANGED);
 or attach as .txt or .zip file
-->
		string decoded = base64_decode(base64image);
		vector<uchar> vectordata(decoded.begin(), decoded.begin() + decoded.size());
		auto mat = imdecode(vectordata, IMREAD_UNCHANGED);


With debug build, the number of rows and columns come out to be less than 1000. With the release built the number of rows/ columns come out to be 100k+ .

**Comment by berak** [comment]
your base64 decoder works on std::string, and cannot handle binary images(containing 0's)

**Comment by berak** [comment]
are you aware, that you cannot put binary data, like images into a std::string ? (decoded.end() will terminate at the 1st `0` found)

**Comment by siddhupiddu** [comment]
I have tested this with decoded.start() + decoded.size() as the second argument as well. std::string does not have a null terminated string like c strings.

**Comment by siddhupiddu** [comment]
After debugging, I can verify that the following function:
inline Mat& Mat::operator = (const Mat& m)
in opencv2/core/mat.hpp (tag 2.4.13) gets a Mat reference that has the right rows, cols and data. However, the returned Mat is all wrong.

**Comment by alalek** [comment]
Please provide a complete example with dumping of interested values (via `printf` or `std::cout`).
Also please add exact expected and received result.

**Comment by siddhupiddu** [comment]
Will do. I think this happens due to a bug introduced by compiler optimization in the copy constructor and '=' overloading function in mat.hpp.

**Comment by alalek** [comment]
In OpenCV 3.x+ there is [move assignment operator](https://github.com/opencv/opencv/blob/b2fa82ffcd14798391312567c554948888cd9503/modules/core/include/opencv2/core/mat.inl.hpp#L1171) also. It is enabled in C++11 compilation mode (enabled by default in MSVS2012+).

**Comment by kinchungwong** [comment]
If this is a compiler bug, please contact the compiler vendor.

I have had found a couple of Visual C++ bugs. Here is what is needed in order to file a bug report with the compiler vendor:

 * Need to have the full source code of the libraries being compiled.
 * Try to create the minimally reproducible example, by extracting bits of source code (from the application as well as the affected library) into a blank project.
 * You also need to know the C++ standard, http://en.cppreference.com/w/ , this is because bugs come from implementations, and implementations are designed to handle the cases specified by the C++ standard. Compiler bugs will have very narrow window of exposure (corner cases).
 * You also need to analyze the disassembly to ascertain the compiled code's behavior.
 * Basically, if you have a minimally reproducible example that is less than 100 lines long, you can submit that to the compiler vendor's bug report system. Keep a copy of all of the intermediate build results, in particular the EXE and the PDB, which are necessary for disassembly analysis.

However, some compiler vendors do not accept bug reports on older versions of compilers. For example, Microsoft considers that if a bug is already fixed on VS2015, it is your responsibility to upgrade to VS2015, not Microsoft's responsibility to fix it in VS2013 (since there will not be any more updates to older compilers).

**Comment by siddhupiddu** [comment]
Yes, it is due to quirks of the visual studio compiler and I was able to fix this issue. I tried the move constructor but ran into the same problem using the move constructor. The constructor was defined in the .hpp file. I did not change the opencv library but rather the .hpp file. I placed a #pragma to turn the optimization off for the move constructor and then everything worked fine. 

Hope this helps any future user and I am closing the issue,
